### PR TITLE
feat(query-bar): Implement multi enum in filter button

### PR DESCRIPTION
**Pull request by ChenCodes**
Changes in this PR:
- Update Condition flowtype `value` and `operator` property to be `Array<string>` instead of `string`
- Use `MultiSelectField` in ValueField.js
- Add multi-enum object into columns array in `fixtures.js` for use in the styleguide
- Cleaned up `fixtures.js` by moving copy/pasted column objects to a constant and share that with other constants
- Update Condition.scss to include error state for MultiSelectField
- Handle handler logic when deleting a date from the DatePicker, ran into a bug earlier where there was an error that said `cannot toString null`
screen shot 2019-03-01 at 2 25 21 pm

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @ChenCodes has signed the CLA. Thanks for the pull request!

File: src/features/query-bar/__tests__/FilterButton-test.js [hunk 1]
```diff
@@ -263,16 +257,50 @@ describe('feature/query-bar/components/filter/FilterButton', () => {
     });
 
     describe('closeOnClickPredicate()', () => {
-        test('Should return true if Apply button was clicked', () => {
+        test('Should return true if Apply button was clicked and value is not empty', () => {
```
Comment (macrotim) [inline src/features/query-bar/__tests__/FilterButton-test.js]: The next two tests deserve to be merged with test.each.

File: src/features/query-bar/components/filter/Condition.js [hunk 2]
```diff
@@ -52,10 +52,10 @@ const Condition = ({
 
     const handleOperatorChange = (option: OptionType) => {
         const { value } = option;
-        onFieldChange(condition, value, OPERATOR);
+        onFieldChange(condition, ([value]: Array<string>), OPERATOR);
```
Comment (macrotim) [inline src/features/query-bar/components/filter/Condition.js]: Ah good fix.

File: src/features/query-bar/components/filter/Condition.js [hunk 3]
```diff
@@ -183,7 +183,7 @@ const Condition = ({
                         isDisabled={false}
                         onChange={handleOperatorChange}
                         options={operatorOptions}
-                        selectedValue={operator}
+                        selectedValue={operator[0]}
```
Comment (macrotim) [inline src/features/query-bar/components/filter/Condition.js]: Why this change?

File: src/features/query-bar/components/filter/FilterButton.js [hunk 4]
```diff
@@ -69,13 +69,13 @@ class FilterButton extends React.Component<Props, State> {
         const { columns } = this.props;
         if (columns) {
             const firstColumn = columns[0];
-            const operator = COLUMN_OPERATORS[firstColumn.type][0].key;
+            const operator = [COLUMN_OPERATORS[firstColumn.type][0].key];
```
Comment (macrotim) [inline src/features/query-bar/components/filter/FilterButton.js]: Is it possible for Condition to support multiple operators? If not, let's go this route.

File: src/features/query-bar/components/filter/FilterButton.js [hunk 5]
```diff
@@ -69,13 +69,13 @@ class FilterButton extends React.Component<Props, State> {
...
 
             return {
                 columnId: firstColumn.id,
                 id: conditionID,
                 operator,
-                value: null,
+                value: [],
```
Comment (macrotim) [inline src/features/query-bar/components/filter/FilterButton.js]: If it's plural, then let's name it that way i.e. `values.`

File: src/features/query-bar/components/filter/ValueField.js [hunk 6]
```diff
@@ -4,20 +4,23 @@ import { FormattedMessage } from 'react-intl';
...
 import { VALUE } from '../../constants';
 import messages from '../../messages';
 
 import '../../styles/Condition.scss';
 
 type Props = {
-    onChange: (value: String) => void,
-    selectedValue?: string | number,
+    onChange: (value: Array<string>) => void,
+    selectedValue: Array<string>,
```
Comment (macrotim) [inline src/features/query-bar/components/filter/ValueField.js]: Let's pluralize.

File: src/features/query-bar/components/filter/ValueField.js [hunk 7]
```diff
@@ -4,20 +4,23 @@ import { FormattedMessage } from 'react-intl';
...
     valueOptions: Array<Object>,
     valueType: string,
 };
 
 const ValueField = ({ onChange, selectedValue, valueOptions, valueType }: Props) => {
+    const isValueSet = selectedValue.length > 0;
```
Comment (macrotim) [inline src/features/query-bar/components/filter/ValueField.js]: Pluralize selectedValue.

File: src/features/query-bar/components/filter/ValueField.js [hunk 8]
```diff
@@ -70,20 +73,32 @@ const ValueField = ({ onChange, selectedValue, valueOptions, valueType }: Props)
                         hideLabel
                         label="Date"
                         name="datepicker"
-                        onChange={e => onChange(e.toString())}
+                        onChange={e => {
+                            return e ? onChange([e.toString()]) : onChange([]);
```
Comment (macrotim) [inline src/features/query-bar/components/filter/ValueField.js]: Could it fire with an empty event? I'm wonder if this ternary op is needed?

File: src/features/query-bar/components/filter/ValueField.js [hunk 9]
```diff
@@ -39,9 +42,9 @@ const ValueField = ({ onChange, selectedValue, valueOptions, valueType }: Props)
                         hideLabel
                         label="Number input"
                         name="number field"
-                        onChange={e => onChange(e.target.value)}
+                        onChange={e => onChange([e.target.value])}
                         placeholder="Enter a number"
-                        value={selectedValue || ''}
+                        value={isValueSet ? selectedValue[0] : null}
```
Comment (macrotim) [inline src/features/query-bar/components/filter/ValueField.js]: This ternary is duplicated. Can it be moved up?

File: src/features/query-bar/components/filter/ValueField.js (hunk 8 above)
Comment (ChenCodes) [reply src/features/query-bar/components/filter/ValueField.js]: It still fires with a null event.

File: src/features/query-bar/components/filter/ValueField.js [hunk 10]
```diff
@@ -20,6 +20,7 @@ type Props = {
```
Comment (ChenCodes) [inline src/features/query-bar/components/filter/ValueField.js]: I committed this change with the wrong message, it should be 'refactor value in ValueField'

File: src/features/query-bar/flowTypes.js [hunk 11]
```diff
@@ -24,6 +24,6 @@ export type ColumnType = {
 export type ConditionType = {
     columnId: string,
     id: number,
-    operator: Array<string>,
-    value: Array<string>,
+    operator: string,
```
Comment (macrotim) [inline src/features/query-bar/flowTypes.js]: Too bad `operator` isn't typed. Something to consider.

File: src/features/query-bar/__tests__/FilterButton-test.js [hunk 12]
```diff
@@ -99,34 +99,34 @@ describe('feature/query-bar/components/filter/FilterButton', () => {
...
                         columnId: '1',
                         id: '5',
                         operator: EQUALS,
-                        value: null,
+                        values: [],
                     },
                 ],
-                property: 0,
-                conditionProperty: 'value',
+                value: ['0'],
```
Comment (macrotim) [inline src/features/query-bar/__tests__/FilterButton-test.js]: Plural?

File: src/features/query-bar/__tests__/FilterButton-test.js [hunk 13]
```diff
@@ -257,38 +257,24 @@ describe('feature/query-bar/components/filter/FilterButton', () => {
...
-            wrapper.instance().setState({
-                conditions,
-            });
-            targetWithClassName.target.className = 'apply-filters-button';
-            const closeOnClickPredicateResult = wrapper.instance().shouldClose(targetWithClassName);
-
-            wrapper.update();
-            expect(closeOnClickPredicateResult).toEqual(true);
-        });
+        const condition = {
```
[the remark this answers is not in the document]
Comment (ChenCodes) [reply src/features/query-bar/__tests__/FilterButton-test.js]: Yes

File: src/features/query-bar/components/filter/Condition.js [hunk 14]
```diff
@@ -23,7 +23,7 @@ type Props = {
     index: number,
     onColumnChange: (condition: ConditionType, columnId: string) => void,
     onConnectorChange: (option: OptionType) => void,
-    onFieldChange: (condition: ConditionType, value: Array<string>, property: string) => void,
+    onFieldChange: (condition: ConditionType, value: string | Array<string>, property: string) => void,
```
[the remark this answers is not in the document]
Comment (ChenCodes) [reply src/features/query-bar/components/filter/Condition.js]: I could break it up into two methods but then we would have two methods that both do about 95% of the same thing:

```
handleOperatorChange = (condition: ConditionType, value: string) => {
        const { conditions } = this.state;
        let newConditionIndex = 0;
        const conditionToUpdate = conditions.find((currentCondition, index) => {
            newConditionIndex = index;
            return currentCondition.id === condition.id;
        });

        const newCondition = {
            ...conditionToUpdate,
            [OPERATOR]: value,
        };

        const newConditions = conditions.slice(0);
        newConditions[newConditionIndex] = newCondition;

        this.setState({
            conditions: newConditions,
        });
    };
```

and 

```
handleValueChange = (condition: ConditionType, values: Array<string>) => {
        const { conditions } = this.state;
        let newConditionIndex = 0;
        const conditionToUpdate = conditions.find((currentCondition, index) => {
            newConditionIndex = index;
            return currentCondition.id === condition.id;
        });

        const newCondition = {
            ...conditionToUpdate,
            [VALUES]: values,
        };

        const newConditions = conditions.slice(0);
        newConditions[newConditionIndex] = newCondition;

        this.setState({
            conditions: newConditions,
        });
    };
```

**Comment by ChenCodes** [comment]
I think Operator gets the point across. Is there a meaning that OperatorKey helps convey that would give the user more context?
